src/noise-models.lisp

```diff
@@ -0,0 +1,262 @@
+
+
+
+;;; A noise model consists of a set of noise rules and readout-povms.
+;;; The noise rules define quantum channels and where they should be
+;;; applied, and the readout-povms describe the noisy readout
+;;; probabilities.
+(defclass noise-model ()
+  ((noise-rules
+    :initarg :noise-rules
+    :accessor noise-rules
+    :initform nil
+    :documentation "A list of noise-rules. Each NOISE-RULE is a NOISE-PREDICATE and a list of kraus maps.")
+   (readout-povms
+    :initarg :readout-povms
+    :accessor readout-povms
+    :initform (make-hash-table :test 'eql)
+    :documentation "A map of qubit to readout noise assignment probabilities."))
+  (:documentation "A NOISE-MODEL is a collection of NOISE-RULES, describing what type of noise should be applied to a program that goes through a qvm, and where that noise should be applied."))
+
+
+(defmethod initialize-instance :after ((nm noise-model) &rest args)
+  ;; Ensure the noise rules are sorted according to priority order
+  (declare (ignore args))
+  (let ((rules (noise-rules nm)))
+    (setf (slot-value nm 'noise-rules) (sort (copy-list rules) #'rule-priority>=))))
+
+
+(defun (setf qubit-povm) (povm nm qubit)
+  "Set the povm for a qubit in this noise model."
+  (check-povm povm)
+  (setf (gethash qubit (readout-povms nm)) povm))
+
+
+(defun make-noise-model (noise-rules)
+  "Returns a noise model with noise-rules."
+  (make-instance 'noise-model :noise-rules noise-rules))
+
+
+(defun add-noise-models (nm1 nm2)
+  "Combines two noise models by concatenating their rules. This way, either rule can be matched by an instruction."
+  (make-noise-model (append (noise-rules nm1) (noise-rules nm2))))
+
+
+(defun multiply-noise-models (nm1 nm2)
+  "Combines two noise models in a way such that both set of rules, or one and not the other could be matched by an instruction."
+  (make-noise-model
+   (loop :for r1 :in (noise-rules nm1)
+         :append (loop :for r2 :in (noise-rules nm2)
+                       :collect (apply #'make-noise-rule
+                                       (predicate-and (noise-predicate r1)
+                                                      (noise-predicate r2))
+                                       (append (operation-elements r1)
+                                               (operation-elements r2)))
+                       :collect (apply #'make-noise-rule
+                                       (predicate-and (noise-predicate r1)                
+                                                      (predicate-not (noise-predicate r2)))
+                                       (operation-elements r1))
+                       :collect (apply #'make-noise-rule
+                                       (predicate-and (noise-predicate r2)                   
+                                                      (predicate-not (noise-predicate r1)))
+                                       (operation-elements r2))))))
+
+
+(defun match-strict-qubits (&rest qubits)
```

Comment:
I would rename these functions `match-...` to `...-matcher`. They don't literally match, but produce a boolean function that matches.

Reply:
`...-predicate`